contracts/upgradeable_contracts/arbitrary_message/MessageDelivery.sol

```diff
@@ -2,22 +2,56 @@ pragma solidity 0.4.24;
 
 import "openzeppelin-solidity/contracts/math/SafeMath.sol";
 import "./BasicAMB.sol";
+import "../../libraries/ArbitraryMessage.sol";
+import "../../libraries/Bytes.sol";
 
 contract MessageDelivery is BasicAMB {
     using SafeMath for uint256;
 
-    function requireToPassMessage(address _contract, bytes _data, uint256 _gas) public {
+    /**
+    * @dev Requests message relay to the opposite network
+    * @param _contract executor address on the other side
+    * @param _data calldata passed to the executor on the other side
+    * @param _gas gas limit used on the other network for executing a message
+    */
+    function requireToPassMessage(address _contract, bytes _data, uint256 _gas) public returns (bytes32) {
         require(_gas >= getMinimumGasUsage(_data) && _gas <= maxGasPerTx());
-        emitEventOnMessageRequest(abi.encodePacked(msg.sender, _contract, _gas, uint8(0x00), _data));
+
+        bytes memory sourceChainId = _sourceChainId();
+        bytes memory destinationChainId = _destinationChainId();
+
+        // 4 bytes - message version
+        // 20 bytes - bridge id
+        // 8 bytes - message nonce
+        bytes32 messageId = MESSAGE_PACKING_VERSION |
+            (keccak256(abi.encodePacked(sourceChainId, address(this))) &
```

Comment:
Bearing other comments in mind the following refactoring could be suggested. 

Consider to introduce a new method `packHeader(address, uint256)`.
Here is quite optimized version of the method:
```solidity
function _packHeader(address _contract, uint256 _gas) view internal returns(bytes header) {
    uint256 sourceChainId = _sourceChainId();
    uint256 sourceChainIdLength = _sourceChainIdLenght();
    uint256 destinationChainId = _destinationChainId();
    uint256 destinationChainIdLength = _destinationChainIdLength();
    
    uint256 mVer = MESSAGE_PACKING_VERSION; //note: it must be 0x00400000, not shifted
    uint256 nonce = _nonce();
    
    bytes32 bridgeId = keccak256(abi.encodePacked(sourceChainId, address(this)));
    // 79 = 4 + 20 + 8 + 20 + 20 + 4 + 1 + 1 + 1
    header = new bytes(79 + sourceChainIdLength + destinationChainIdLength);

    assembly {
        let header_ptr := add(header, 32)
        mstore(header_ptr, mVer)
        mstore(add(header_ptr, 4), shl(32, bridgeId))
        mstore(add(header_ptr, 20), shl(192, nonce))
        mstore(add(header_ptr, 32), shl(96, caller))
        mstore(add(header_ptr, 52), shl(96, _contract))
        mstore(add(header_ptr, 72), shl(224, _gas))
        mstore8(add(header_ptr, 76), sourceChainIdLength)
        mstore8(add(header_ptr, 77), destinationChainIdLength)
        mstore8(add(header_ptr, 78), 0x0)
        mstore(add(header_ptr, 79), shl(sub(256, shl(3, sourceChainIdLength)), 
                                        sourceChainId)) // shl(3, sourceChainIdLength) == sourceChainIdLength * 8
        mstore(add(add(header_ptr, 79), sourceChainIdLength), shl(sub(256, shl(3, destinationChainIdLength)), 
                                                                  destinationChainId)) // shl(3, destinationChainIdLength) == destinationChainIdLength * 8
    }
}
```
The returned value of this method will be encoded with `_data` in order to get full message.

It will allow to keep chain ids and their lengths as integers (rather than set of bytes) that should simplify an incoming message checking and setting the values through the initializer and the setter.

Reply:
As far as I can see, we can also switch the order of header packing (from last field to first). This will allow as to get rid of unnecessary `shl` operations.

Reply:
cool! Very clever!